lib/network/http_middlewares.go

```diff
@@ -35,3 +43,102 @@ func RecoverMiddleware(logger log15.Logger) mux.MiddlewareFunc {
 		})
 	}
 }
+
+func rateLimitReachedHandler(w http.ResponseWriter, r *http.Request) {
+	httputils.WriteJSONError(w, errors.ErrorTooManyRequests)
+}
+
+func rateLimitErrorHandler(w http.ResponseWriter, r *http.Request, err error) {
+	httputils.WriteJSONError(w, errors.ErrorHTTPServerError.Clone().SetData("error", err))
+}
+
+// RateLimitMiddleware throttles the incoming requests; if `Limit` is 0, there
+// will be no limit.
+func RateLimitMiddleware(logger logging.Logger, rule common.RateLimitRule) mux.MiddlewareFunc {
+	if logger == nil {
+		logger = log
+	}
+
+	store := memory.NewStoreWithOptions(
+		limiter.StoreOptions{
+			CleanUpInterval: time.Duration(2) * time.Minute,
+		},
+	)
+
+	var defaultMiddleware *stdlib.Middleware
+	if rule.Default.Limit > 0 {
+		defaultMiddleware = stdlib.NewMiddleware(
+			limiter.New(store, rule.Default),
+			stdlib.WithForwardHeader(true),
+			stdlib.WithErrorHandler(rateLimitErrorHandler),
+			stdlib.WithLimitReachedHandler(rateLimitReachedHandler),
+		)
+	}
+
+	middlewaresByIP := map[ /* ip address */ string]*stdlib.Middleware{}
+	for ip, rate := range rule.ByIPAddress {
+		var m *stdlib.Middleware
+		if rate.Limit > 0 {
+			m = stdlib.NewMiddleware(
+				limiter.New(store, rate),
+				stdlib.WithForwardHeader(true),
+				stdlib.WithErrorHandler(rateLimitErrorHandler),
+				stdlib.WithLimitReachedHandler(rateLimitReachedHandler),
+			)
+		}
+		middlewaresByIP[ip] = m
+	}
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			ip := limiter.GetIPKey(r, true)
+
+			// from localhost, no rate limit
+			if strings.HasPrefix(ip, "127.0.") {
+				w.Header().Add("X-RateLimit-Limit", "")
+				w.Header().Add("X-RateLimit-Remaining", "")
+				w.Header().Add("X-RateLimit-Reset", "")
+
+				next.ServeHTTP(w, r)
+				return
+			}
+
+			var middleware *stdlib.Middleware
+
+			if len(middlewaresByIP) < 1 {
+				middleware = defaultMiddleware
+			} else { // find middleware by ip
+				var found bool
+				if middleware, found = middlewaresByIP[ip]; !found {
```

Comment:
I have a question, Within this using, `middlwaresByIP` is thread-safty when multiple gorountes run for HTTP handling?

Reply:
@anarcher `middlwaresByIP` will be created before running `http.Handler`, so reading `middlwaresByIP` will be safe in goroutine :)

Reply:
Oh, I got it. :+1:  If we don't update this map in running times, I agree that this map has thread safety :-)